### ISSUE TITLE
Direct modex broken when dstore GDS used

### ISSUE THREAD

**Issue by rhc54** [pmix/pmix#586]
Please see https://github.com/pmix/pmix/pull/585 for an explanation of the problem.

**Comment by rhc54** [comment]
Transferring this description here so it doesn't get lost:

When requesting modex data without it having already been collected, the client issues a request to the server for data "put" by the given rank. Since the server always returns all data for that proc, we don't pass the key being requested. Thus, the server operates with a NULL key for the request.

When a proc "commits" its modex data, the server stores it in different locations based on its scope. Local scope data goes into the GDS component for that peer, but remote scope data is stored under the server's GDS component. This is done to allow proper response to remote vs local procs - i.e., local procs don't see remote scope data, and vice versa.

The GDS/hash component is working correctly. When a request for data arrives, we look in the appropriate place based on whether the requested proc is local or remote. If we find it, we return that data. If we don't find it, then we create a tracker and wait for it to appear.

The GDS/ds12 component, however, is NOT working correctly. Because a NULL key is provided, it only looks to see if the job-level data is available, and then immediately returns with SUCCESS and the shared memory location. This is okay IF the requested proc has already "committed" its data, but causes the client proc to generate an error if the requested proc has NOT provided its data yet.

This is seen when running the following test in the test/simple directory:
```
$ export PMIX_SIMPDMODEX_ASYNC=1
$ ./simptest -n 2 -e ./simpdmodex
```
@artpol84 @karasevb I think this is probably a blocker for v2.1 - can you please verify?

**Comment by artpol84** [comment]
Indeed it looks like a blocker. We will take a look next week.

**Comment by rhc54** [comment]
It looks like this is going to hold up OMPI v3.1 release (as it would break your latest SLURM plugin enhancement), so please keep us updated

**Comment by artpol84** [comment]
Yes, we will.

@karasevb this is a high prio!

**Comment by artpol84** [comment]
@rhc54 Can you provide more details on the connection with Slurm plugin?

**Comment by artpol84** [comment]
Also I do you know if this is something that was a problem from the day 1st or this is related to the recent cross-version implementation?

**Comment by rhc54** [comment]
I think this may have been around for awhile - at least, I observed it prior to the recent commit. It impacts your direct modex support in the SLURM plugin because of the race condition - if someone asks for modex info prior to the peer having committed it, then the request will return an error.

**Comment by artpol84** [comment]
I guess that in case of the OMPI this shouldn't be the case as we are doing barrier after commit if direct modex is on. What do you think?

**Comment by rhc54** [comment]
Only in the case where the async_mpi_init flag isn't given - otherwise, you do indeed hit it. We discovered the problem when doing the scaling test.

**Comment by artpol84** [comment]
Ok, now it makes sense why we missed it.

**Comment by karasevb** [comment]
I have found the problem, work on fix.

**Comment by artpol84** [comment]
@karasevb did a deeper analysis of the problem and it looks like it's not only `GDS/dstore`. Also `GDS/hash` is affected but in more rare cases.

If the patch below is applied to the simple test we are getting hang with `GDS/hash` as well. What this patch does - it emulates the situation where no local keys was submitted and the client requests non-existing. In this case server can't find the key it will assume that it wasn't yet commited and will start waiting on it. But this will never happen. I suspect that the same case can occur with remote keys, @karasevb is going to confirm that experimentaly.

So our solution for both problems is to track all local processes commits and check additionally if commit was seen - we assume that the key should be present and if not found - just respond immediately. @rhc54 if you are ok with this - Boris is going to implement this tomorrow.

```patch
diff --git a/test/simple/simpdmodex.c b/test/simple/simpdmodex.c
index b12afdb..59cf8fb 100644
--- a/test/simple/simpdmodex.c
+++ b/test/simple/simpdmodex.c
@@ -125,6 +125,7 @@ int main(int argc, char **argv)
     pmix_output(0, "Client %s:%d universe size %d", myproc.nspace, myproc.rank, nprocs);
 
     /* put a few values */
+    /*
     (void)asprintf(&tmp, "%s-%d-internal", myproc.nspace, myproc.rank);
     value.type = PMIX_UINT32;
     value.data.uint32 = 1234;
@@ -140,6 +141,7 @@ int main(int argc, char **argv)
         pmix_output(0, "Client ns %s rank %d: PMIx_Put internal failed: %d", myproc.nspace, myproc.rank, rc);
         goto done;
     }
+    */
 
     (void)asprintf(&tmp, "%s-%d-remote", myproc.nspace, myproc.rank);
     value.type = PMIX_STRING;
```

**Comment by rhc54** [comment]
Yeah, if they ask for something that we never receive, then we will wait forever - unless they provided the timeout key (which probably still needs to be implemented). I don't consider that a bug as we are doing what they asked us to do (i.e., wait for the data to arrive). We do need to implement the timeout key.

I think the problem that needs to be solved is the one where we report success that the data is available, but it actually isn't. I'd suggest we just deal with that for now, and I'll add the timeout key separately.

**Comment by artpol84** [comment]
What I proposed will fix both problems. 
We want to track commits. This seems like an easiest way to do that.
We have an option to implement that internally in `GDS/dstore` or do that at the pmix server level and address hash issues as well.

**Comment by artpol84** [comment]
As far as I understand non-existing keys are possible and some implementation may use that as a flag (though not the best way to do that). So I consider this as a corner-case bug.

**Comment by rhc54** [comment]
I'm not sure I understand, so let me try to grok it. What you are proposing is to immediately return an error if at least one call to PMIx_Commit was done by a process, and we don't find the requested key? If so, then wouldn't this will break the folks that rely on multiple calls to commit/get unless they guarantee to always put a fence in-between...which would be a new requirement?

**Comment by artpol84** [comment]
"I think the problem that needs to be solved is the one where we report success that the data is available, but it actually isn't. "

One would treat the hang I described with the simptest as reporting local key as "existing" but not yet arrived.

**Comment by rhc54** [comment]
Not sure I understand - how do you know the key "exists" if it has not yet arrived?

**Comment by artpol84** [comment]
simptest with the patch I attached is hanging. It hangs because server is waiting for the local key to arrive (pending request). But this key will never arrive.

**Comment by artpol84** [comment]
" If so, then wouldn't this will break the folks that rely on multiple calls to commit/get unless they guarantee to always put a fence in-between...which would be a new requirement?"

Then I have a question. How do we distinguish non-existing and existing keys? Do you say that if a non-existing key is requested you supposed to pay a timeout penalty?
I thought that OMPI relies on non-existing keys through "optional" flag. Or somehow jobinfo is different from commited keys?

**Comment by rhc54** [comment]
"simptest with the patch I attached is hanging. It hangs because server is waiting for the local key to arrive (pending request). But this key will never arrive."

Correct - and that is the correct behavior. It isn't a bug. We have no way of knowing if that key will ever arrive. The user asked us to fetch it and didn't tell us to give up after some time, so we should wait forever for it.

"Then I have a question. How do we distinguish non-existing and existing keys? Do you say that if a non-existing key is requested you supposed to pay a timeout penalty?"

Yes - like I said, there is no way we can know that this key will never exist

"I thought that OMPI relies on non-existing keys through "optional" flag. Or somehow jobinfo is different from commited keys?"

Yes, OMPI does use "optional", and that is fine. If the "optional" is given, then the client lib needs to respond immediately as that is what they asked us to do. Likewise, "immediate" means that the PMIx server should check to see if it is present, and return immediately with the answer.

If neither is given, and no timeout value is provided, then we must wait forever to respond...as that is what they are asking us to do.

**Comment by artpol84** [comment]
Yes I forgot about this optional flag.
Here is one more thing that is not clear to me. Server treats all rank keys as a blob. If rank has commited at least once `GDS/hash` has the blob and will respond with it even though the desired key is located in say second commit. How this aligned with multiple-commit usage?

**Comment by rhc54** [comment]
Hmmm...good point. ```gds/hash``` is supposed to return all received blobs, not just the first one, but that does beg the question of why we are responding if the requested key hasn't arrived yet.

Seems to me that the local server should include in its request the desired key and any timeout instructions. The remote server has access to the actual key-value pairs (it constructs the blob upon request), and so it can wait for the key to arrive before responding. Means a little more logic in the server, but nothing too horrid.

Make sense?

**Comment by artpol84** [comment]
One more issue that won't be solved by this is a key overwriting. You won't have a way to know if the key was updated and the client has no way to indicate what version it wants.

**Comment by rhc54** [comment]
True - we can note that in the standard (i.e., overwriting of keys must be followed by a fence prior to get). I don't know how else to handle it since we don't know when it might occur. Thankfully, that's a really rare use-case.

**Comment by artpol84** [comment]
To summarize what we have discussed here.
If we are going to follow this route:
"...Seems to me that the local server should include in its request the desired key and any timeout instructions. The remote server has access to the actual key-value pairs (it constructs the blob upon request), and so it can wait for the key to arrive before responding. Means a little more logic in the server, but nothing too horrid..."

Then dstore as it is now seems to work fine. Can we close this issue in favor of new one where we will implement the solution that you have proposed ?

**Comment by rhc54** [comment]
I think we still need to fix the behavior where dstore reports the key as being present when it isn't. In the current code, if someone asks for modex info prior to the target proc doing a commit, dstore only looks to see if the job-level data is present and then tells the client that all data is present. At the very least, we need it to (a) see if any modex data for that proc has been received, (b) request it if not, and (c) not respond until at least some modex data has been returned.

We can then work the logic of what constitutes a completed return - but at least we won't be getting a false positive response.

**Comment by artpol84** [comment]
According to what you proposed above we will also pass a specific key. This will be an "indication" of modex.  

`dstore` only communicates with local processes. All remote requests will go through `hash`. So the scope is reduced to local process requesting data from another local process. 

There can be following cases if we pass the key along with request:
* target local proc submitted the data, then response will be correct.
* target local proc hasn't yet commited. In this case we won't find the key and will wait for it to arrive (or timeout).

In both cases it should work well.

**Comment by artpol84** [comment]
I'm trying to converge to avoid extra work.

**Comment by artpol84** [comment]
I just talked to @karasevb and he raised another valid question regarding the approach that you proposed.
What if client requested multiple keys? As it is working right now we turn any key request into a rank-blob request and if client requests N keys from the same rank only one message will go to the server.
According to what have been said above we can't rely on the assumption that if the first requested key was found - others will be found as well.

I think that we should:
* implement a temporal solution in the dstore to track if the data from the key was commited. @karasevb please implement our original solution with an array of flags for the local(remote?) ranks.
* we need to discuss this and come up with the consistent solution as all what have been discussed so far has its week sides.

**Comment by rhc54** [comment]
Understood. Let's follow the progression and see what actually occurs so we can best plan out the right solution. The problem we have right now is that local procs ask for modex data from other local procs (e.g., for shared memory locations) and get a "not found" error if the requested proc has not yet committed its info.

What happens today is:

1. the client proc does a "get" on another local peer. We first check the internal storage (where it will not be found because we didn't store it). Since the key is not NULL and doesn't start with "pmix", we do a gds/fetch against the local server's GDS component - which is the dstore.

1. the local client dstore component checks to see if the job-level data for the specified nspace is available. Since that was given to us when we called PMIx_Init, the answer is "yes" and so the dstore then looks for the given key and doesn't find it (since the requested proc has not yet committed it). Thus, the gds/fetch returns "not found" as the specific key isn't there.

1. the local client then generates a request to the server for the specified data, and a message is sent to the server telling it the nspace/rank of the target proc.

1. the server knows the nspace because it was registered prior to launching the clients. Likewise, we can assume that all local clients have called PMIx_Init and are therefore registered. This causes the server to call the _satisfy_request function.

1. _satisfy_request sees that this is a local request, and so it uses the GDS component for this nspace - in this case, it will again select the dstore component. We then do a gds/fetch on the dstore component asking for data for the specified peer. I am not entirely clear what happens down in there, but it appears we return a "not found" error.

1. the server then sends this back to the local client. Specifically, it doesn't "hold" the request, nor does it pass the request upwards as the request was for data from a local proc.

So what needs to happen is that somewhere in this procedure, someone needs to "hold" the request until the key is committed. It seems to me like it is best done in the server get routine and not in the gds/dstore itself.

I'll respond to your new note separately.

**Comment by rhc54** [comment]
"According to what have been said above we can't rely on the assumption that if the first requested key was found - others will be found as well."

Correct - we have to pass all the requested keys and wait for them to be committed before completing the request.

"implement a temporal solution in the dstore to track if the data from the key was commited. @karasevb please implement our original solution with an array of flags for the local(remote?) ranks"

I'm afraid I don't understand what you are proposing when you say a "temporal" solution. I would rather not create another array, nor do I see any reason for it. We already have a peer object and should be able to track if the peer has committed something in there. Why do we need something new?

"we need to discuss this and come up with the consistent solution as all what have been discussed so far has its week sides."

No issue with doing so - there will be no perfect solution, you know. Everything involves trade-offs. We already designed this with a timeout capability (that needs to be implemented), and chose to default to waiting forever as that is the behavior expected by current MPI and other libraries.

The thing to remember here is that we are talking about _corner-cases_ now - the direct modex behavior when data has not yet been committed is the unusual case, and we don't want to design everything around it. We first need to ensure that the typical case is being handled correctly. Per my prior note, it isn't being handled correctly today, and that is what I'd like to see us fix prior to release.

The corner-cases can come as we go forward - I'm not immediately concerned about covering all of them.

**Comment by artpol84** [comment]
"
"implement a temporal solution in the dstore to track if the data from the key was commited. @karasevb please implement our original solution with an array of flags for the local(remote?) ranks"

I'm afraid I don't understand what you are proposing when you say a "temporal" solution. I would rather not create another array, nor do I see any reason for it. We already have a peer object and should be able to track if the peer has committed something in there. Why do we need something new?
"
I originally suggested to implement it that way (calling it server level):
https://github.com/pmix/pmix/issues/586#issuecomment-351113475
But had impression that you didn't liked a solution. I suggested to respond immediately if we found that commit went in. As it is now - this is the best we can do. It can be done quickly and can work into closest release. I don't care where to keep this information and as I've said in https://github.com/pmix/pmix/issues/586#issuecomment-351113475 I prefer to have it at the server level (peers structure) rather than at dstore level. So no issues with implementing it using peers info.

**Comment by rhc54** [comment]
I fear I am getting lost in all the kerfluffle. Let me take a crack at solving the problem I've identified, and then let's circle around again.

Given that people are going to start disappearing in another 24 hours for the holidays, I suspect this is going to shift into Jan anyway. So let me play a little with it once I get a couple of other things out of the way.

**Comment by artpol84** [comment]
I suggest to have a call. It will be faster. Can we do that tomorrow?
I agree that this chat is getting messy.

**Comment by rhc54** [comment]
A call is fine - will have to be in the afternoon.

**Comment by artpol84** [comment]
I have some idea on how to solve it with client involvement. As you've said missing modex is a rare case, so if it will take multiple messages to resolve rare case - not an issue.

**Comment by artpol84** [comment]
Ok, will ping you tomorrow.

**Comment by artpol84** [comment]
I want to fix the idea here so I will not forget. 

Notes:
* It is important for performance that we don't send request for each key, otherwise it will be similar to what is happening with Slurm/PMI2 plugin and this is a bottleneck.
* The client requests keys one by one - this is how PMIx_Get is structured. So I believe it will be a problem to requests multiple keys:
  * from one side we want to stack keys together to reduce amount of messages (though we don't want a big message too).
  * from the other hand we don't want to wait for other keys to arrive before sending the request as will significantly increase the latency.

I see 2 possible 2-step protocols to preserve the performance for the common case and still allow a corner case though slower.


Version 1:
At the server side we have a counter, that holds a **rank-info-version** (this is a node-local thing!) of the data that is known for each rank in the GDS.
1. Client issues a regular request for the whole rank blob without specifying the key. 
2. Server checks the commit flag. If flag is `false` - we are waiting for the data to arrive from the target peer. If the flag is `true` - we respond immediately appending **rank-info-version** for client reference.
3. At the client side:
  * if there was a race condition client may find the key at this time and all be OK.
  * otherwise - client issues another kind of request (wait request) that will tell server that this client wants to wait until **rank-info-version** is incremented and then respond. 
4. If server gets wait request - it checks if **rank-info-version** hasn't already changed:
  * If yes - responds to a client (goto step 3)
  * If no - wait for a new data to arrive and respond then (goto Step 3).

Version 2:
At the server side we have a flag for each local peer that reflects whether something has beet commited from this peer at least once.
1. First client issues a regular request for the whole rank blob without specifying the key. 
2. Server checks the commit flag. If flag is `false` - we are waiting for the data to arrive from the target peer. If the flag is `true` - we respond immediately.
3. At the client side:
  * if there was a race condition client may find the key at this time and all be OK.
  * otherwise - client issues another kind of request (clarification request) putting all of the keys that are pending for this rank. 
4. If server gets clarification request - it address each key in that request and classifies them into 2 groups
  * available keys - for them a clarification response is formed
  * not available keys - those are inserted into a pending queue.

I'm in favor of the version 1 :)

**Comment by artpol84** [comment]
Makes sense to note here that this protocol suggests data aging that was discussed in https://github.com/pmix/pmix/issues/119. We (well, maybe I) keep returning to this question.

**Comment by artpol84** [comment]
This should be fixed in #613. 
@rhc54 please confirm.

**Comment by rhc54** [comment]
Agreed - thanks guys!
